examples/vision/image_classification_with_vision_transformer.py

```diff
@@ -0,0 +1,294 @@
+    def __init__(self, num_patches, projection_dims):
+        super(PatchEncoder, self).__init__()
+        self.projection = layers.Dense(units=projection_dims)
+        self.position_embedding = layers.Embedding(
+            input_dim=num_patches, output_dim=projection_dims
+        )
+
+    def call(self, patch):
+        positions = tf.range(start=0, limit=num_patches, delta=1)
+        encoded = self.projection(patch) + self.position_embedding(positions)
+        return encoded
+
+
+"""
+### Build the ViT model
+
+The ViT model consists of multiple layers of the Transformer block,
+which uses the `layers.MultiHeadAttention` layer as a self-attention mechanism
+for the sequence of patches. The Transformer blocks produces a
+`[batch_size, num_patches, projection_dim]` tensor, which is processed via an
+MLP head with softmax to produce the final class probabilities output.
+
+Unlike the technique described in the [paper](https://arxiv.org/abs/2010.11929),
+which prepends a learnable embedding to the sequence of encoded patches to serve
+as the image representation, the outputs of the final Transformer block are
+aggregated with `layers.GlobalAveragePooling1D()` and then used as the image
+representation input to the MLP head.
+"""
+
+
+def create_vit_classifier():
+    inputs = layers.Input(shape=input_shape)
+    # Augment data.
+    augmented = data_augmentation(inputs)
+    # Create patches.
+    patches = Patches(patch_size)(augmented)
+    # Encode patches.
+    encoded_patches = PatchEncoder(num_patches, projection_dims)(patches)
+
+    # Create multiple layers of the Transformer block.
+    for _ in range(transfomer_layers):
+        # Layer normalization 1.
+        x1 = layers.LayerNormalization(epsilon=1e-6)(encoded_patches)
+        # Create a multi-headed attention layer.
+        attention_output = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=projection_dims, dropout=dropout_rate
+        )(x1, x1)
+        # Skip connection 1.
+        x2 = layers.Add()([attention_output, encoded_patches])
+        # Layer normalization 2.
+        x3 = layers.LayerNormalization(epsilon=1e-6)(x2)
+        # MLP.
+        x3 = MLP(transformer_hidden_units, dropout_rate)(x3)
+        # Skip connection 2.
+        encoded_patches = layers.Add()([x3, x2])
+
+    # Create a [batch_size, projection_dims] tensor.
+    representation = layers.GlobalAveragePooling1D()(encoded_patches)
+    representation = layers.LayerNormalization(epsilon=1e-6)(representation)
+    # Create MLP.
+    features = MLP(hidden_units, dropout_rate)(representation)
+    # Create softmax output.
+    outputs = layers.Dense(num_classes, activation="softmax")(features)
+    # Create the Keras model.
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    return model
+
+
+vit_classifier = create_vit_classifier()
+history = run_experiment(vit_classifier)
+
+"""
+After 100 epochs, the ViT classification model achieves more than 72% top 5
```

Comment:
Was the ViT model benchmarked on CIFAR100 by someone before? What accuracy did it achieve?

Reply:
It was benchmarked in the paper indeed. However, I didn't try to replicate their results, since they performed pre-training on the JFT-300M dataset before fine-tuning on CIFAR-100. In addition, they used much larger values for the hyperparams (e.g 12 to 32 transformer layers, compared to 4 in this example, 12-16 attention heads compared to 4 in this example, projection size 768-1280 compared to 64 in this example).